python/src/utils.h

```diff
@@ -44,6 +44,23 @@ inline array to_array_with_accessor(nb::object obj) {
   }
 }
 
+inline bool is_convertable_to_array(const ScalarOrArray& v) {
+  // Checks if the value can be converted to an array (or is already an
+  // mlx array)
+  if (auto pv = std::get_if<nb::bool_>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<nb::int_>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<nb::float_>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<std::complex<float>>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<nb::object>(&v); pv) {
+    return nb::hasattr(*pv, "__mlx_array__") || nb::isinstance<array>(*pv);
+  }
```

Comment:
You had a different version earlier which just pulled out an object if it existed. So if it's not an `nb::oject` then it's ok (int, bool, float, complex) and if it is then check if its an array or has the attribute? Maybe we should do that instead?

Reply:
It's the same line that is present here:
```
else if (auto pv = std::get_if<nb::object>(&v); pv) {
    return nb::hasattr(*pv, "__mlx_array__") || nb::isinstance<array>(*pv);
  }
```
The function that we had earlier (`isMlxCoreArray`) checked this line. I think right now it's better and we may use this utility function further in more operations (where we need to check if the operands are valid and can be converted to `mlx.core.array`).

Reply:
Ok just thinking about it from an efficiency standpoint. Can you not do?

```
if (auto pv = std::get_if<nb::object>(&v); pv) {
    return nb::isinstance<array>(*pv) || nb::hasattr(*pv, "__mlx_array__");
} else {
    return true;
}
```

Also it might be minor but I would swap the `hasattr` and the `isinstance` since the former will be much much more common (and also probably a lot less expensive).

Reply:
I swapped the `hasattr` and `isinstance`. Also, why should we add `else { return true; }`? Because if it's not each of the options mentioned above (bool/int/float/complex/array) it is not convertible to array and thus, we should simply return `False`.

Reply:
please check the lates commit and let me know about what you think should be changed

Reply:
Also, I think we need a better name for this function. Because, instances like `list` and `tuple` are actually convertible to `mlx.core.array`. They are not __comparable__ to arrays. Don't you agree? This name might be a little misleading in further developments